File: api-gateway/src/main/java/com/kjj/apigateway/config/AddIdToHeaderFilter.java
package com.kjj.apigateway.config;

import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

import java.util.UUID;

@Component
@Slf4j
@Order(1)
public class AddIdToHeaderFilter extends AbstractGatewayFilterFactory {
    @Override
    public GatewayFilter apply(Object config) {
        return (exchange, chain) -> {
            // request id 생성
            String requestId = UUID.randomUUID().toString();

            // header에 request id 추가
            exchange.getRequest().mutate().headers(httpHeaders -> httpHeaders.add("X-Request-ID", requestId));

            // logback에서 사용하기 위해 MDC에 request id 추가
            MDC.put("requestId", requestId);

            return chain.filter(exchange);
        };
    }
}

File: gateway/src/main/kotlin/com/kjj/gateway/config/AddIdToHeaderFilter.kt
package com.kjj.gateway.config

import org.slf4j.MDC
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import java.util.*

@Component
@Order(1)
class AddIdToHeaderFilter: AbstractGatewayFilterFactory<Any>() {
    override fun apply(config: Any): GatewayFilter {
        return GatewayFilter { exchange, chain ->
            val requestId = UUID.randomUUID().toString()

            exchange.request.mutate().headers { httpHeaders -> httpHeaders.add("X-Request-ID", requestId) }

            MDC.put("requestId", requestId)

            chain.filter(exchange)
        }
    }
}